### PR TITLE
feat: generalize `finrank_add_finrank_quotient` to PIDs

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Finiteness.lean [hunk 1]
```diff
@@ -644,6 +644,16 @@ theorem equiv [Finite R M] (e : M ≃ₗ[R] N) : Finite R N :=
 
 instance ulift [Finite R M] : Finite R (ULift M) := equiv ULift.moduleEquiv.symm
 
+theorem iff_FG {N : Submodule R M} : Module.Finite R N ↔ N.FG := Module.finite_def.trans (fg_top _)
```
Comment (alreadydone) [inline Mathlib/RingTheory/Finiteness.lean:647]:
```suggestion
theorem iff_fg {N : Submodule R M} : Module.Finite R N ↔ N.FG := Module.finite_def.trans (fg_top _)
```

File: Mathlib/Algebra/Module/PID.lean [hunk 2]
```diff
@@ -280,3 +280,67 @@ theorem equiv_free_prod_directSum [h' : Module.Finite R N] :
...
+  intro hc
+  obtain ⟨x, rfl⟩ := Submodule.mkQ_surjective _ x
+  rw [← LinearMap.map_smul, Submodule.mkQ_apply, Submodule.Quotient.mk_eq_zero] at hcx
+  obtain ⟨n, hn⟩ := hcx
+  simp only [Submodule.mkQ_apply, Submodule.Quotient.mk_eq_zero, Submonoid.mk_smul, exists_prop]
+  refine ⟨n * ⟨c, mem_nonZeroDivisors_of_ne_zero hc⟩, ?_⟩
+  simpa [Submonoid.smul_def, smul_smul] using hn
+
+instance [IsDomain R] [IsPrincipalIdealRing R] : Module.Free R (M ⧸ Submodule.torsion R M) :=
+  Module.free_of_finite_type_torsion_free'
```
Comment (alreadydone) [inline Mathlib/Algebra/Module/PID.lean:297]: Some build failures here and below

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean [hunk 3]
```diff
@@ -436,7 +436,98 @@ instance Module.Finite.span_singleton (x : V) : Module.Finite R (R ∙ x) :=
...
 
-variable {K R}
+variable {K R M}
+variable [AddCommGroup M] [Module R M] [Module.Finite R M]
+
+lemma LinearIndependent.finset_toSet (s : Finset M) : LinearIndependent R ((↑) : (↑s : Set M) → M) ↔
+    LinearIndependent R ((↑) : s → M) := Iff.rfl
+
+lemma FiniteDimensional.finrank_add_finrank_quotient_le (N : Submodule R M) :
+    finrank R N + finrank R (M ⧸ N) ≤ finrank R M := by
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:446]: Just letting you know: I'm trying to prove the `rank` version of this and I've established a [key lemma](https://github.com/leanprover-community/mathlib4/compare/master...rank_quotient_add_or_prod#diff-23b5582f59f20ba2c8878d54747dac6dcabb32114e16e5f5d92dd53652cabe5eR854-R857) in another branch using your prior work in #8842.

**Comment by erdOne** [comment]
Closing in favor of #9298

**Pull request closed without merging**